### PR TITLE
Adjust the cpp-extensions when publishing

### PR DESCRIPTION
**Pull request by lmcbout**
### Description
Adjust package.json to publish latest or next
Add script to use the next | latest for Theia dependencies
Prevent git to auto-commit the code when publishing

#### Note: to make sure you don't publish, move temporarily the file in your home folder before running the test
    mv <home folder>/.npmrc .npmrcBK
    When testing, this will generate an error for publishing, 
    but all the process will be executed

In the main repo of the cpp-extensions, run the following command:
Publish latest
   `yarn run publish:latest`
Expected result
 ```
- You will be asked for a version (Patch, Minor, Major...)select one and pres" Enter"
- It will show you what the Changes willl be and you need to provide an answer: Yes | No 
- Selecting "Yes" will initiate  a publish process. If your credential are set properly, then the application will be published.

Look in lerna.json,  ==> the version you specify
Look in all package.json: ==> the specified version is applied to 
   @theia/cpp-debug 
   @theia/cortex-debug. 
All other @theia package should show latest
```
Publish next
   `yarn run publish:next`
Expected result
  `Similar result, except the "@theia/... : latest will be "@theia/...: next-<version identification>"`

Signed-off-by: Jacques Bouthillier <jacques.bouthillier@ericsson.com>

### CODE REVIEW

File: lerna.json [hunk 1]
```diff
@@ -1,11 +1,14 @@
 {
   "lerna": "2.4.0",
-  "version": "0.0.0",
+  "version": "0.7.0",
```
Comment (vince-fugnitto) [inline lerna.json]: Based on discussions with Marc, we should keep the version as `0.0.0`
(and align the `cpp-debug` and `cortex-debug` to `0.0.0`)
This means that when we publish (the first latest), `lerna` will step the versions properly at `0.1.0`.

cc @marcdumais-work

File: scripts/theia-version.sh [hunk 2]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
```
Comment (vince-fugnitto) [inline scripts/theia-version.sh]: Was this code copied from somewhere?

Comment (lmcbout) [reply scripts/theia-version.sh]: Script from theia-go-extensions

File: lerna.json (hunk 1 above)
Comment (lmcbout) [reply lerna.json]: Agree Done

File: scripts/theia-version.sh (hunk 2 above)
Comment (lmcbout) [reply scripts/theia-version.sh]: I would be happy to put the licence, but there is non in the theia-go-extensions

Comment (marcdumais-work) [reply scripts/theia-version.sh]: Note: as per the Foundation workflow .pdf, a CQ is still required because the original code is not under Foundation control. It should be a formality to have it accepted though.

File: scripts/theia-version.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
```
Comment (vince-fugnitto) [inline scripts/theia-version.js]: I don't believe this is the correct license.
The original license is only `EPL-2.0`.
See [here](https://github.com/TypeFox/theia-xtext-sprotty-example/blob/master/LICENSE)

File: scripts/theia-version.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+ ********************************************************************************/
+/********************************************************************************
+ * Copyright (c) 2019 TypeFox and others
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
```
Comment (vince-fugnitto) [inline scripts/theia-version.js]: The proper license would simply be `EPL-2.0` without the mention of `GNU Classpath Exception`.
https://www.eclipse.org/legal/epl-2.0/faq.php#h.q72cnghf29k0

For example (from the Eclipse documentation):
```
/*********************************************************************
* Copyright (c) {date} {owner} [and others]
*
* This program and the accompanying materials are made
* available under the terms of the Eclipse Public License 2.0
* which is available at https://www.eclipse.org/legal/epl-2.0/
*
* SPDX-License-Identifier: EPL-2.0
**********************************************************************/
```

File: lerna.json [hunk 5]
```diff
@@ -6,6 +6,9 @@
   "command": {
     "run": {
       "stream": true
+    },
+    "publish": {
+      "skipGit": true
```
Comment (vince-fugnitto) [inline lerna.json]: For my own information, what is `skipGit` used for and why was it necessary as a change?
Perhaps this is something useful to include as part of the commit body as it doesn't look trivial in relation to the PR.

**Comment by vince-fugnitto** [comment]
The CI failures are related to ffmpeg and should be addressed by the PR #55 which benefits from the upstream fix to `@theia/electron`.

**Comment by lmcbout** [comment]
I raised a CQ ( CQ20877  ), but I also raised a bug in Bugzilla ( https://bugs.eclipse.org/bugs/show_bug.cgi?id=551459 ) since I cannot upload the file for the CQ, need to wait until I can upload the file to continue with  the CQ

File: lerna.json (hunk 5 above)
Comment (lmcbout) [reply lerna.json]: As discuss with Marc, we should be able to look at the modified files before committing to git. With this flag, when running the publish command, it does not create automatically a commit and push.
This is the same behavior as the main theia branch.

**Comment by lmcbout** [comment]
@marcdumais-work @vince-fugnitto 
[CQ](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20877) approved, need an approval before pushing or do you have any more comments?

**Comment by vince-fugnitto** [comment]
> @marcdumais-work @vince-fugnitto
> [CQ](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20877) approved, need an approval before pushing or do you have any more comments?

A PR approval is needed before merging :)
I'd like if the PR #55 is merged beforehand, so this PR can be rebased and it's CI can pass.
Although, it might not be 100% necessary.

**Comment by marcdumais-work** [comment]
@vince-fugnitto as discussed yesterday, can you try building the extensions without a yarn.lock and confirm it works? If so we can seriously consider just removing it, since it will cause maintenance headaches, whenever we release.

**Comment by marcdumais-work** [comment]
> @vince-fugnitto as discussed yesterday, can you try building the extensions without a yarn.lock and confirm it works? If so we can seriously consider just removing it, since it will cause maintenance headaches, whenever we release.

Let's wait for this before merging this PR here.

**Comment by vince-fugnitto** [comment]
@lmcbout you can now rebase the PR against `master` as #55 was merged :)

**Comment by lmcbout** [comment]
Re-based and is currently running Travis

**Comment by vince-fugnitto** [comment]
@lmcbout like @marcdumais-work mentioned in the meeting, can you rephrase your commit title and message (to properly express the changes/why it was changed). 

Also, can you express if the PR should be tested, how we can successfully test it?

**Comment by lmcbout** [comment]
@vince-fugnitto Just change the commit message, this is what is going on with this PR.

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto Just change the commit message, this is what is going on with this PR.

I guess the idea is:

> Updated the `package.json` scripts `publish:latest` and `publish:next` which are used to publish both the latest and next versions of the extensions present in the repository, respectfully. The added `theia-version.js` automates the update of dependencies to their appropriate versions in order to publish successfully, while the `publish.git` config for `lerna.js` ensures that no commit is automatically created or pushed.

Also, I mentioned on how it can be tested, can you include it perhaps in the description of the PR above? Or should we just merge?

**Comment by lmcbout** [comment]
@vince-fugnitto 
- Description updated to show how to test
- Commit message as suggested

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto
> * Description updated to show how to test

I'm not sure it's the proper way to test, won't it attempt to actually publish?

**Comment by lmcbout** [comment]
@vince-fugnitto  I added a note in "How to test" to make sure you don't publish to npm when testing

File: scripts/theia-version.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ ********************************************************************************/
+
+ /* copied code from https://github.com/TypeFox/theia-xtext-sprotty-example/blob/master/theia/theia-version.js#L1-26 */
+
+const shell = require('shelljs')
```
Comment (vince-fugnitto) [inline scripts/theia-version.js]: When running the command `yarn run publish:latest` I get the following error:

```
yarn publish:latest
yarn run v1.17.3
$ node scripts/theia-version.js latest && lerna publish && yarn publish:check
internal/modules/cjs/loader.js:584
    throw err;
    ^

Error: Cannot find module 'shelljs'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:582:15)
    at Function.Module._load (internal/modules/cjs/loader.js:508:25)
    at Module.require (internal/modules/cjs/loader.js:637:17)
    at require (internal/modules/cjs/helpers.js:22:18)
    at Object.<anonymous> (/Users/vincentfugnitto/workspace/theia-cpp-extensions/scripts/theia-version.js:28:15)
```

A `devDependency` to **shelljs** is likely required.

Comment (vince-fugnitto) [reply scripts/theia-version.js]: Since **shelljs** is not a dependency present under the main repository, I believe it will
require a new CQ.

File: scripts/theia-version.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
+/********************************************************************************
+ * Copyright (C) 2019 Ericsson and others.
+ *
```
Comment (marcdumais-work) [inline scripts/theia-version.js]: I am not sure if adding our license in this case is the right thing to do. We used the pattern before in the main repo, but it was for MIT-licensed copied code, which allows re-licensing. I do not know that we can do the same with EPL2. I think the safe thing to do is just keep the original license header and do our own changes under that license.

File: scripts/theia-version.js (hunk 6 above)
Comment (marcdumais-work) [reply scripts/theia-version.js]: I guess it should be added in package.json? It would be a dev-dependency, which has a different procedure for CQ: 
https://github.com/eclipse-theia/theia/wiki/Registering-CQs#case-new-or-updated-testbuild-dependency-eg-devdependencies

Comment (vince-fugnitto) [reply scripts/theia-version.js]: Yes exactly, a `devDependency` under the root's `package.json`.
Great that the procedure is different!

File: scripts/theia-version.js (hunk 7 above)
Comment (lmcbout) [reply scripts/theia-version.js]: Remove Ericsson copyright and kept the TypeFox one.
Uddated the year of the copyright

File: scripts/theia-version.js (hunk 6 above)
Comment (lmcbout) [reply scripts/theia-version.js]: Registered a CQ for shelljs
https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20969

**Comment by marcdumais-work** [comment]
I think we'll need to try a release or two to see if the strategy adopted here will work or need changing. As it is, I think we'll need to proceed something like:
- Along with the main repo release (e.g. 0.12.0), release a `latest` version of the extensions of this repo here
- create a release commit that matches what we released
- immediately add a commit on top where all dependencies are switched back to `next` and the extensions of this repo are versioned back to the release version (e.g. 0.12.0) both in their `package.json` and that of the example applications
